helper/resource/state_shim.go

```diff
@@ -186,3 +186,37 @@ func shimmedAttributes(instance *states.ResourceInstanceObjectSrc, res *schema.R
 
 	return instanceState.Attributes, nil
 }
+
+func shimLegacyState(legacy *terraform.State) (*states.State, error) {
+	state, err := terraform.ShimLegacyState(legacy)
+	if err != nil {
+		return nil, err
+	}
+
+	if state.HasResources() {
+		for _, module := range state.Modules {
+			for name, resource := range module.Resources {
+				module.Resources[name].ProviderConfig.Provider = addrs.NewDefaultProvider(resource.Addr.Resource.ImpliedProvider())
```

Comment:
Should this line use `addrs.ImpliedProviderForUnqualifiedType`? I imagine it's unlikely that we'll have test states with a builtin provider, so it's not _really_ necessary, but it might be safer.